parcagpu/parcagpu.go

```diff
@@ -0,0 +1,175 @@
+package parcagpu // import "go.opentelemetry.io/ebpf-profiler/parcagpu"
+
+import (
+	"context"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+	"unsafe"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/perf"
+	log "github.com/sirupsen/logrus"
+
+	"go.opentelemetry.io/ebpf-profiler/host"
+	"go.opentelemetry.io/ebpf-profiler/libpf"
+	"go.opentelemetry.io/ebpf-profiler/support"
+)
+
+type mapKey struct {
+	pid uint32
+	id  uint32
+}
+
+type gpuTraceFixer struct {
+	mu                  sync.Mutex
+	timesAwaitingTraces map[mapKey]kernelTimingEvent
+	tracesAwaitingTimes map[mapKey]*host.Trace
+}
+
+func (p *gpuTraceFixer) addTrace(trace *host.Trace) *host.Trace {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	key := mapKey{uint32(trace.PID), trace.ParcaGPUTraceID}
+	ev, ok := p.timesAwaitingTraces[key]
+	if ok {
+		delete(p.timesAwaitingTraces, key)
+		trace.OffTime = int64(ev.end - ev.start)
+		prepTrace(trace, &ev)
+		return trace
+	}
+	p.tracesAwaitingTimes[key] = trace
+	return nil
+}
+
+func (p *gpuTraceFixer) addTime(key mapKey, ev *kernelTimingEvent) *host.Trace {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	trace, ok := p.tracesAwaitingTimes[key]
+	if ok {
+		delete(p.tracesAwaitingTimes, key)
+		prepTrace(trace, ev)
+		return trace
+	}
+	p.timesAwaitingTraces[key] = *ev
+	return nil
+}
+
+// uprobes aren't perfect and we may miss matching timing to trace at attach boundary
+// so clear them if they get too big.
+func (p *gpuTraceFixer) clear() {
```

Comment:
This should have a name that indicates it doesn't _always_ clear, perhaps `maybeClear` ?